File: python-app/routes/clientes.py
from flask import Blueprint, request, jsonify
from models import Cliente
from app import db

clientes_bp = Blueprint('clientes', __name__)

@clientes_bp.route('/', methods=['GET'])
def get_clientes():
    """Listar todos os clientes ativos"""
    clientes = Cliente.query.filter_by(ativo=True).order_by(Cliente.nome).all()
    return jsonify([c.to_dict() for c in clientes])

@clientes_bp.route('/<uuid>', methods=['GET'])
def get_cliente(uuid):
    """Buscar cliente por UUID"""
    cliente = Cliente.query.filter_by(uuid=uuid).first()
    if not cliente:
        return jsonify({'error': 'Cliente não encontrado'}), 404
    return jsonify(cliente.to_dict())

@clientes_bp.route('/search', methods=['GET'])
def search_clientes():
    """Buscar clientes por termo"""
    termo = request.args.get('termo', '')
    clientes = Cliente.query.filter(
        Cliente.ativo == True,
        db.or_(
            Cliente.nome.ilike(f'%{termo}%'),
            Cliente.cpf.ilike(f'%{termo}%'),
            Cliente.telefone.ilike(f'%{termo}%')
        )
    ).order_by(Cliente.nome).all()
    return jsonify([c.to_dict() for c in clientes])

@clientes_bp.route('/', methods=['POST'])
def create_cliente():
    """Criar novo cliente"""
    data = request.get_json()
    
    if not data or not all(k in data for k in ['nome', 'cpf', 'telefone']):
        return jsonify({'error': 'Campos obrigatórios: nome, cpf, telefone'}), 400
    
    # Verificar se CPF já existe
    if Cliente.query.filter_by(cpf=data['cpf']).first():
        return jsonify({'error': 'CPF já cadastrado'}), 400
    
    cliente = Cliente(
        nome=data['nome'],
        cpf=data['cpf'],
        telefone=data['telefone'],
        email=data.get('email')
    )
    
    db.session.add(cliente)
    db.session.commit()
    
    return jsonify({'success': True, 'uuid': cliente.uuid}), 201

@clientes_bp.route('/<uuid>', methods=['PUT'])
def update_cliente(uuid):
    """Atualizar cliente"""
    cliente = Cliente.query.filter_by(uuid=uuid).first()
    if not cliente:
        return jsonify({'error': 'Cliente não encontrado'}), 404
    
    data = request.get_json()
    if not data:
        return jsonify({'error': 'Nenhum dado fornecido'}), 400
    
    # Atualizar campos
    if 'nome' in data:
        cliente.nome = data['nome']
    if 'cpf' in data:
        # Verificar se CPF já existe em outro cliente
        outro = Cliente.query.filter(Cliente.cpf == data['cpf'], Cliente.uuid != uuid).first()
        if outro:
            return jsonify({'error': 'CPF já cadastrado'}), 400
        cliente.cpf = data['cpf']
    if 'telefone' in data:
        cliente.telefone = data['telefone']
    if 'email' in data:
        cliente.email = data['email']
    
    db.session.commit()
    
    return jsonify({'success': True})

@clientes_bp.route('/<uuid>', methods=['DELETE'])
def delete_cliente(uuid):
    """Excluir cliente (soft delete)"""
    cliente = Cliente.query.filter_by(uuid=uuid).first()
    if not cliente:
        return jsonify({'error': 'Cliente não encontrado'}), 404
    
    cliente.ativo = False
    db.session.commit()
    
    return jsonify({'success': True})

File: php-app/api/clientes.php
<?php
// ==========================================
// API REST - CLIENTES
// ==========================================
require_once '../config/database.php';
setJsonHeaders();

$method = $_SERVER['REQUEST_METHOD'];
$action = $_GET['action'] ?? '';

try {
    $db = getDB();
    
    switch ($method) {
        case 'GET':
            if ($action === 'all') {
                // Listar todos os clientes ativos
                $stmt = $db->prepare("SELECT * FROM clientes WHERE ativo = 1 ORDER BY nome");
                $stmt->execute();
                jsonResponse($stmt->fetchAll());
                
            } elseif ($action === 'get' && isset($_GET['uuid'])) {
                // Buscar cliente por UUID
                $stmt = $db->prepare("SELECT * FROM clientes WHERE uuid = ?");
                $stmt->execute([$_GET['uuid']]);
                $cliente = $stmt->fetch();
                
                if ($cliente) {
                    jsonResponse($cliente);
                } else {
                    jsonResponse(['error' => 'Cliente não encontrado'], 404);
                }
                
            } elseif ($action === 'search' && isset($_GET['termo'])) {
                // Buscar clientes por termo
                $termo = '%' . $_GET['termo'] . '%';
                $stmt = $db->prepare("
                    SELECT * FROM clientes 
                    WHERE ativo = 1 AND (nome LIKE ? OR cpf LIKE ? OR telefone LIKE ?)
                    ORDER BY nome
                ");
                $stmt->execute([$termo, $termo, $termo]);
                jsonResponse($stmt->fetchAll());
                
            } else {
                jsonResponse(['error' => 'Ação inválida'], 400);
            }
            break;
            
        case 'POST':
            if ($action === 'create') {
                // Criar novo cliente
                $data = json_decode(file_get_contents('php://input'), true);
                
                if (!isset($data['nome']) || !isset($data['cpf']) || !isset($data['telefone'])) {
                    jsonResponse(['error' => 'Campos obrigatórios: nome, cpf, telefone'], 400);
                }
                
                $uuid = gerarUUID();
                $stmt = $db->prepare("
                    INSERT INTO clientes (uuid, nome, cpf, telefone, email, ativo)
                    VALUES (?, ?, ?, ?, ?, 1)
                ");
                $stmt->execute([
                    $uuid,
                    $data['nome'],
                    $data['cpf'],
                    $data['telefone'],
                    $data['email'] ?? null
                ]);
                
                jsonResponse(['success' => true, 'uuid' => $uuid], 201);
                
            } else {
                jsonResponse(['error' => 'Ação inválida'], 400);
            }
            break;
            
        case 'PUT':
            if ($action === 'update' && isset($_GET['uuid'])) {
                // Atualizar cliente
                $data = json_decode(file_get_contents('php://input'), true);
                $uuid = $_GET['uuid'];
                
                $campos = [];
                $valores = [];
                
                if (isset($data['nome'])) {
                    $campos[] = 'nome = ?';
                    $valores[] = $data['nome'];
                }
                if (isset($data['cpf'])) {
                    $campos[] = 'cpf = ?';
                    $valores[] = $data['cpf'];
                }
                if (isset($data['telefone'])) {
                    $campos[] = 'telefone = ?';
                    $valores[] = $data['telefone'];
                }
                if (isset($data['email'])) {
                    $campos[] = 'email = ?';
                    $valores[] = $data['email'];
                }
                
                if (empty($campos)) {
                    jsonResponse(['error' => 'Nenhum campo para atualizar'], 400);
                }
                
                $valores[] = $uuid;
                $sql = "UPDATE clientes SET " . implode(', ', $campos) . " WHERE uuid = ?";
                $stmt = $db->prepare($sql);
                $stmt->execute($valores);
                
                jsonResponse(['success' => true]);
                
            } else {
                jsonResponse(['error' => 'Ação inválida'], 400);
            }
            break;
            
        case 'DELETE':
            if ($action === 'delete' && isset($_GET['uuid'])) {
                // Soft delete (desativar)
                $uuid = $_GET['uuid'];
                $stmt = $db->prepare("UPDATE clientes SET ativo = 0 WHERE uuid = ?");
                $stmt->execute([$uuid]);
                
                jsonResponse(['success' => true]);
                
            } else {
                jsonResponse(['error' => 'Ação inválida'], 400);
            }
            break;
            
        default:
            jsonResponse(['error' => 'Método não permitido'], 405);
    }
    
} catch (Exception $e) {
    jsonResponse(['error' => $e->getMessage()], 500);
}
?>
